authorization-common/src/main/java/com/sequenceiq/authorization/service/EnvironmentBasedAuthorizationProvider.java

```diff
@@ -34,61 +35,61 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(EnvironmentBasedAuthorizationProvider.class);
 
     @Inject
-    private CommonPermissionCheckingUtils commonPermissionCheckingUtils;
+    private Map<AuthorizationResourceType, AuthorizationEnvironmentCrnListProvider> environmentCrnListProviderMap;
+
+    @Inject
+    private Map<AuthorizationResourceType, AuthorizationEnvironmentCrnProvider> environmentCrnProviderMap;
 
     public Optional<AuthorizationRule> getAuthorizations(String resourceCrn, AuthorizationResourceAction action) {
-        ResourcePropertyProvider resourceBasedCrnProvider = commonPermissionCheckingUtils.getResourceBasedCrnProvider(action);
-        if (resourceBasedCrnProvider == null) {
-            LOGGER.error("There is no resource based crn provider implemented for action {} against resource type {}, " +
-                    "thus authorization is failing automatically.", action, Crn.safeFromString(resourceCrn).getResourceType().name());
-            throw new AccessDeniedException(String.format("Action %s is not supported over resource %s, thus access is denied",
-                    action.getRight(), resourceCrn));
-        }
-        Optional<String> environmentCrn = resourceBasedCrnProvider.getEnvironmentCrnByResourceCrn(resourceCrn);
-        if (environmentCrn.isPresent()) {
-            return Optional.of(new HasRightOnAny(action, List.of(environmentCrn.get(), resourceCrn)));
+        if (action.getAuthorizationResourceType().isHierarchicalAuthorizationNeeded()) {
+            AuthorizationEnvironmentCrnProvider environmentCrnProvider = environmentCrnProviderMap.get(action.getAuthorizationResourceType());
+            if (environmentCrnProvider == null) {
+                LOGGER.error("There is no resource based crn provider implemented for action {} against resource type {}, " +
+                        "thus authorization is failing automatically.", action, Crn.safeFromString(resourceCrn).getResourceType().name());
+                throw new AccessDeniedException(String.format("Action %s is not supported over resource %s, thus access is denied",
+                        action.getRight(), resourceCrn));
+            }
+            return Optional.of(new HasRightOnAny(action, List.of(environmentCrnProvider.getEnvironmentCrnByResourceCrn(resourceCrn), resourceCrn)));
```

Comment:
What if for some reason we don't have the environment crn? For example the datalake is still present but the environment was deleted? We had such instances in mow-dev. Then the user won't be able to delete the datalake because we constantly blow up right?

Reply:
So you mean we should keep the result of  getEnvironmentCrn method optional? Anyway I do not think it is a valid use case in prod and also we should make sure authorization happens against environment too so we cannot just skip it if environment crn is not present.